### DISCUSSION TITLE
Not able to increase read size beyond 128K in Fuse3

### DISCUSSION THREAD

**Discussion by bhansalivikas** [libfuse/libfuse#845] [General]
Environment:
- Operating System: Ubuntu 20.04.5 LTS
-  Kernel: Linux 5.15.0-1036-azure
-  Fuse lib : libfuse3

Issue:
During init I am setting max_write to 1MB, max_read to 1MB and mount was done with "-o max_read=1048576". As per documentation on kernel beyond 4.20 read calls shall come with 1MB buffers but in my tests I always get 128K read calls. Writes are coming with 1MB buffer but not able to tune reads with such high buffer.

On trying "-o max_pages=256" it gives an error about not supported option. 

Is there anyway to tune read calls to use 1MB buffer instead of 128KB.

**Comment by mharvey-jt** [comment]
On at least the rhel 8.6 kernel, the kfuse code caps the setting of readahead at 128kiB. The kernel itself is able to prefetch up to 2MiB.
To override the prefetch, poke the buffer size you want in kB into `/sys/class/bdi/<device>/read_ahead_kb`, eg:
```
echo 1024 > /sys/class/bdi/0:433/read_ahead_kb
```
getting the correct value for `device` is a little fiddly, it's obtainable by expanding  the `st_dev` field from at `struct stat` with the `gnu_dev_major` and `gnu_dev_minor` macros:

```
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <sys/types.h>

void main(int argc, char**argv) {
struct stat st;
stat(argv[1], &st);
printf("%d:%d\n",  gnu_dev_major(st.st_dev), gnu_dev_minor(st.st_dev));
}
```

**Comment by bhansalivikas** [comment]
This is about tweaking the read-ahead size while I want to upgrade my read size from 128K to 1MB if possible. Most libfuse document says that this will be by default but in any of my read calls I do not see it going beyond 128KB.

**Comment by trapexit** [comment]
You have to use direct_io or set the filesystem's readahead.

**Comment by bsbernd** [comment]
This is somehow an FAQ, I'm pretty sure I had already written this somewhere before. You can do it, as root and after the mount, like this

`dev=$(stat --format="%Hd:%Ld"  /tmp/mnt/) && echo "1024" >/sys/class/bdi/${dev}/read_ahead_kb`

**Reply by trapexit** [reply]
Yes, in my fs I spawn a thread during init to set the readahead and then it exits. This is necessary because the call to set the readhead will block till the init finishes. Clunky but works.

**Comment by bsbernd** [comment]
@vibhansa-msft You cannot tweak *FUSE* read-size, if the  kernel  reading part does not request more than your read-ahead window. If you don't believe us, try to tweak read-ahead *down* to 4K and then read your file - read size will now be 4K.
